projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,201 @@
+    data->channel_readings[channel][0] = 0;
+    data->channel_readings[channel][1] = 0;
+    data->channel_callback[channel] = NULL;
+    data->callback_context[channel] = NULL;
+  }
+
+  status_ok_or_return(prv_reset(data));
+  status_ok_or_return(
+      prv_setup_register(data, ADS1015_ADDRESS_POINTER_CONFIG, CONFIG_REGISTER_MSB_0,
+                         CONFIG_REGISTER_LSB));  // sets the starting channel to channel 0
+  status_ok_or_return(prv_setup_register(data, ADS1015_ADDRESS_POINTER_LO_THRESH,
+                                         ADS1015_LO_THRESH_REGISTER_MSB,
+                                         ADS1015_LO_THRESH_REGISTER_LSB));
+  status_ok_or_return(prv_setup_register(data, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+
+  // ALERT/RDY continues to require a pullup resistor.
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  status_ok_or_return(gpio_init_pin(data->ready_pin, &gpio_settings));
+  return STATUS_CODE_OK;
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+// It also registers the interrupt handler on ALRT/RDY pin.
+StatusCode ads1015_configure_channel(ADS1015Data *data, ADS1015Channel channel, bool enable,
+                                     ADS1015Callback callback, void *context) {
+  if (!data) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  // if no channels were already enabled, set to given channel
+  if (!data->channel_enable[0] && !data->channel_enable[1] && !data->channel_enable[2] &&
+      !data->channel_enable[3]) {
+    status_ok_or_return(prv_set_channel(data, channel));
+    data->current_channel = channel;
+  }
+
+  data->channel_enable[channel] = enable;
+  data->channel_callback[channel] = callback;
+  data->callback_context[channel] = context;
+
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  StatusCode code = gpio_it_register_interrupt(data->ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
+                                               prv_interrupt_handler, data);
+  prv_check_status(code);
+  status_ok_or_return(code);
+  return STATUS_CODE_OK;
+}
+
+// reads raw 12 bit conversion results which are expressed in two's complement format
+StatusCode ads1015_read_raw(ADS1015Data *data, ADS1015Channel channel, uint16_t *reading) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  *reading =
+      ((data->channel_readings[channel][0] << 8) | (data->channel_readings[channel][1])) >> 4;
+  // data->channel_readings[channel][1];
+  return STATUS_CODE_OK;
+}
+
+// sign extension for two's complement of numbers in n(< 16) bit register to 16 bit register
+static uint16_t prv_sign_ext_16bits(uint8_t num_bits, uint16_t twos_comp_num) {
```

Comment:
If you have a 12-bit signed integer in an int16, you can shift it to the left and then back to sign-extend it. See [this post](https://stackoverflow.com/questions/5814072/sign-extend-a-nine-bit-number-in-c). Technically, signed right shift is undefined according to the C spec, but GCC seems to implement it as an arithmetic right shift that sign extends. [another post](https://stackoverflow.com/questions/7622/are-the-shift-operators-arithmetic-or-logical-in-c)

By the way, you could've done this in one line even without relying on the arithmetic right shift. Your logic is a little more complex than it needs to be.